Framework/DataObjects/inc/MantidDataObjects/WorkspaceCreation.h

```diff
@@ -158,13 +149,39 @@ std::unique_ptr<T> create(const P &parent, const IndexArg &indexArg,
       ws = detail::createConcreteHelper<T>();
     }
   }
+  return ws;
+}
 
+MANTID_DATAOBJECTS_DLL void
+initializeFromParent(const API::MatrixWorkspace &parent,
+                     API::MatrixWorkspace &ws);
+}
+
+template <class T, class P, class IndexArg, class HistArg,
+          class = typename std::enable_if<
+              std::is_base_of<API::MatrixWorkspace, P>::value>::type>
+std::unique_ptr<T> create(const P &parent, const IndexArg &indexArg,
+                          const HistArg &histArg) {
+  std::unique_ptr<T> ws = detail::createUninitialized<T, P>(parent);
   ws->initialize(indexArg, HistogramData::Histogram(histArg));
   detail::initializeFromParent(parent, *ws);
 
   return ws;
 }
 
+template <class T, class P, class IndexArg,
+          class = typename std::enable_if<
+              std::is_base_of<API::MatrixWorkspace, P>::value>::type>
+std::unique_ptr<T> create(const P &parent, const IndexArg &indexArg,
+                          const HistogramData::Histogram &histogram) {
```

Comment:
`EventWorkspace::initialize()` won't accept histograms with non-NULL y or e data. Check the new unit test `test_create_event_from_event()`: it will throw at `create()` when using the original implementation above.
I added this variant to fix the problem. However, as you note, it erroneously  strips the data from histogram workspaces.
Maybe this variant (and the other one below) could be merged to the corresponding `histArg` accepting `create()`:
```c++
template <class T, class P, class IndexArg, class HistArg,
          class = typename std::enable_if<
              std::is_base_of<API::MatrixWorkspace, P>::value>::type>
std::unique_ptr<T> create(const P &parent, const IndexArg &indexArg,
                          const HistArg &histArg) {
  std::unique_ptr<T> ws = detail::createUninitialized<T, P>(parent);
  HistogramData::Histogram templateHist(histArg);
  if (std::is_base_of<DataObjects::EventWorkspace, T>::value) {
    templateHisto = detail::stripData(templateHist);
  }  
  ws->initialize(indexArg, templateHist);
  detail::initializeFromParent(parent, *ws);
  return ws;
}
```
This way we would get rid of the variants and handle EventWorkspaces gracefully. Or am I missing something?